src/milia/api/async_export.cljs

```diff
@@ -6,69 +6,114 @@
+  [dataset-id job-id
+   & [{:keys [on-error on-export-url
+              is-filtered-dataview? millis]
+       :or {:millis 1000}}]]
   (let [done-polling? (atom false)]
     (go
-     (while (not @done-polling?)
-       (let [job-suffix (str "export_async.json?job_uuid=" job-id)
-             job-url (make-url (if is-filtered-dataview? "dataviews" "forms")
-                               dataset-id job-suffix)
-             response (:body (<! (parse-http :get job-url)))]
-         (when-let [export-url (:export_url response)]
-           (on-export-url export-url)
-           (reset! done-polling? true))
-         (<! (timeout millis)))))))
+      (while (not @done-polling?)
+        (let [job-suffix (str "export_async.json?job_uuid=" job-id)
+              job-url    (-> (if is-filtered-dataview?
+                               "dataviews" "forms")
+                           (make-url dataset-id job-suffix))
+              response   (<! (parse-http :get job-url))]
+          ;; Never use `on-job-id` here b/c `on-job-id` should only be
+          ;; triggered once in `trigger-async-export!` where it starts
+          ;; `monitor-async-export!` itself
+          (->> {:on-stop       #(reset! done-polling? true)
+                :on-error      on-error
+                :on-export-url on-export-url}
+            (handle-response response))
+          (<! (timeout millis)))))))
 
 (def export-option-keys
   ["meta" "data_id" "group_delimiter" "do_not_split_select_multiples"
    "remove_group_name" "_version"])
+
 (def export-option-values
-  [:meta-id :data-id :remove-group-name? :do-not-split-multi-selects?
-   :group-delimiter :version])
+  [:meta-id :data-id :group-delimiter :do-not-split-multi-selects?
+   :remove-group-name? :version])
 
 (defn- add-param [key value] (when value (str "&" key "=" value)))
 
 (defn build-export-suffix
   "Build the export options string to pass to the Ona API."
-  [fmt options]
-  (apply str (concat ["export_async.json?format=" fmt]
-                     (map add-param
-                          export-option-keys
-                          (select-values options export-option-values)))))
+  [data-format export-options]
+  (->> export-options
+    ((apply juxt export-option-values))
+    (map add-param export-option-keys)
+    (concat ["export_async.json?format=" data-format])
+    (apply str)))
 
 (defn- trigger-async-export!
   "Triggers async export and watches it via polling.
    Fires on-job-id callback on receving :job_uuid from server, then monitors
    job via polling. On receiving :export_url from server, on-export-url fired."
-  ([dataset-id fmt on-job-id on-export-url]
-   (trigger-async-export! dataset-id fmt on-job-id on-export-url nil))
-  ([dataset-id fmt on-job-id on-export-url
-    {:keys [is-filtered-dataview?] :as options}]
+  ([dataset-id
+    & [{:keys [is-filtered-dataview? data-format export-options
+               ;; callbacks
+               on-job-id on-export-url on-error]}]]
```

Comment:
isn't the `on-job-id` arg always being replaced by the `on-job-id` symbol in the `let` binding (line 100)? If this is intended, there's no need to extract `on-job-id` in the args.